test/generator/define_extern_opencl_aottest.cpp

```diff
@@ -0,0 +1,199 @@
+extern "C" int halide_acquire_cl_context(void *user_context, cl_context *ctx, cl_command_queue *q, bool create = true);
+extern "C" int halide_release_cl_context(void *user_context);
+
+namespace {
+
+cl_program ocl_program;
+
+cl_int init_extern_program() {
+    cl_int error;
+
+    cl_context ocl_ctx = nullptr;
+    cl_command_queue ocl_q = nullptr;
+    int halide_error = halide_acquire_cl_context(nullptr, &ocl_ctx, &ocl_q);
+    if (halide_error != 0) {
+        printf("halide_acquire_cl_context failed (%d).\n", halide_error);
+        return halide_error;
+    }
+
+    const char *ocl_source = "__kernel void add42(__global const int *in, __global int *out) { out[get_global_id(0)] = in[get_global_id(0)] + 42; }";
+    const char *sources[] = { ocl_source };
+    ocl_program = clCreateProgramWithSource(ocl_ctx, 1, &sources[0], NULL, &error);
+    if (error != CL_SUCCESS) {
+        halide_release_cl_context(nullptr);
+        printf("clCreateProgramWithSource failed (%d).\n", error);
+        return error;
+    }
+
+    const cl_uint maxDevices = 4;
+    cl_device_id devices[maxDevices];
+    size_t actual_size = 0;
+    error = clGetContextInfo(ocl_ctx, CL_CONTEXT_DEVICES, sizeof(devices), devices, &actual_size);
+    if (error != CL_SUCCESS) {
+        printf("clGetContextInfo failed (%d).\n", error);
+        halide_release_cl_context(nullptr);
+        return error;
+    }
+
+    error = clBuildProgram(ocl_program, actual_size / sizeof(devices[0]), devices, NULL, NULL, NULL);
+
+    halide_release_cl_context(nullptr);
+
+    if (error != CL_SUCCESS) {
+        if (error == CL_BUILD_PROGRAM_FAILURE) {
+            size_t msg_size;
+            clGetProgramBuildInfo(ocl_program, devices[0], CL_PROGRAM_BUILD_LOG, 0, NULL, &msg_size);
+            char *msg = (char *)malloc(msg_size);
+            clGetProgramBuildInfo(ocl_program, devices[0], CL_PROGRAM_BUILD_LOG, msg_size, msg, NULL);
+            printf("clBuildProgram failed. Error message %s\n", msg);
+        } else {
+            printf("clBuildProgram failed (%d).\n", error);
+        }
+        clReleaseProgram(ocl_program);
+        return error;
+    }
+
+    return error;
+}
+
+void destroy_extern_program() {
+    clReleaseProgram(ocl_program);
+}
+
+}
+
+extern "C" int32_t gpu_input(halide_buffer_t *input, halide_buffer_t *output) {
+    if (input->host == nullptr) {
+        printf("gpu_input: Bounds query for output size %d\n", output->dim[0].extent);
+        input->type = output->type;
+        input->dimensions = 1;
+        input->dim[0] = output->dim[0];
+        return 0;
+    }
+
+    printf("gpu_input: Called to compute on size %d\n", input->dim[0].extent);
+    assert(output->device != 0);
+
+    halide_copy_to_device(nullptr, input, halide_opencl_device_interface());
+
+    cl_int error;
+    cl_context ocl_ctx = nullptr;
+    cl_command_queue ocl_q = nullptr;
+    assert(halide_acquire_cl_context(nullptr, &ocl_ctx, &ocl_q) == 0);
```

Comment:
I think the call should not be in the assert. Asserts probably don't get compiled out in our tests, but it's still a bit sketchy.